File: Design-Patterns-Using-Cpp/Decorator/DecoratorPatternWindow.cpp
#include <iostream>

// Window (Component)
class Window {
public:
    virtual void draw() const {
        std::cout << "SimpleWindow draw." << std::endl;
    }

    virtual ~Window() = default;
};

// SimpleWindow (ConcreteComponent)
class SimpleWindow : public Window {
public:
    void draw() const override {
        std::cout << "SimpleWindow draw." << std::endl;
    }
};

// Decorator (Decorator)
class Decorator : public Window {
protected:
    Window* component;

public:
    Decorator(Window* component) : component(component) {}

    void draw() const override {
        component->draw();
    }

    ~Decorator() override {
        delete component;
    }
};

// VerticalScrollBarDecorator (ConcreteDecorator)
class VerticalScrollBarDecorator : public Decorator {
public:
    VerticalScrollBarDecorator(Window* component) : Decorator(component) {}

    void draw() const override {
        Decorator::draw();
        std::cout << "VerticalScrollBarDecorator draw." << std::endl;
    }
};

// HorizontalScrollBarDecorator (ConcreteDecorator)
class HorizontalScrollBarDecorator : public Decorator {
public:
    HorizontalScrollBarDecorator(Window* component) : Decorator(component) {}

    void draw() const override {
        Decorator::draw();
        std::cout << "HorizontalScrollBarDecorator draw." << std::endl;
    }
};

// Client code
int main() {
    Window* component = new SimpleWindow();
    Window* decorator1 = new VerticalScrollBarDecorator(component);
    Window* decorator2 = new HorizontalScrollBarDecorator(decorator1);

    decorator2->draw();

    // Clean up dynamically allocated memory
    delete decorator2;

    return 0;
}

/*
SimpleWindow draw.
VerticalScrollBarDecorator draw.
HorizontalScrollBarDecorator draw.
*/

File: Design-Patterns-Using-Go/Decorator/DecoratorPatternWindow.go
package main

import "fmt"

// Window is the interface for windows
type Window interface {
	Draw()
}

// SimpleWindow is a simple concrete implementation of Window
type SimpleWindow struct{}

// Draw implements the Draw method for SimpleWindow
func (w *SimpleWindow) Draw() {
	fmt.Println("SimpleWindow draw.")
}

// Decorator is the interface for window decorators
type Decorator interface {
	Window
}

// BaseDecorator provides default implementation for the Decorator interface
type BaseDecorator struct {
	component Window
}

// VerticalScrollBarDecorator is a concrete decorator for adding a vertical scrollbar to a window
type VerticalScrollBarDecorator struct {
	BaseDecorator
}

// Draw implements the Draw method for VerticalScrollBarDecorator
func (d *VerticalScrollBarDecorator) Draw() {
	d.component.Draw()
	fmt.Println("VerticalScrollBarDecorator draw.")
}

// HorizontalScrollBarDecorator is a concrete decorator for adding a horizontal scrollbar to a window
type HorizontalScrollBarDecorator struct {
	BaseDecorator
}

// Draw implements the Draw method for HorizontalScrollBarDecorator
func (d *HorizontalScrollBarDecorator) Draw() {
	d.component.Draw()
	fmt.Println("HorizontalScrollBarDecorator draw.")
}

// Client Code
func main() {
	component := &SimpleWindow{}
	decorator1 := &VerticalScrollBarDecorator{BaseDecorator{component}}
	decorator2 := &HorizontalScrollBarDecorator{BaseDecorator{decorator1}}
	decorator2.Draw()
}

/*
SimpleWindow draw.
VerticalScrollBarDecorator draw.
HorizontalScrollBarDecorator draw.
*/